crates/sui-source-validation/src/lib.rs

```diff
@@ -396,3 +474,273 @@ fn local_modules(
+            Some(toolchain_version) => {
+                println!(
+                    "{} {package} compiler @ {}",
+                    "REQUIRE".bold().green(),
+                    toolchain_version.compiler_version.yellow(),
+                );
+                package_version_map.insert(*package, toolchain_version);
+            }
+        }
+    }
+
+    let mut units = vec![];
+    let mut package_compiled: HashSet<Symbol> = HashSet::new();
+    // Iterate over compiled units, and determine whether they need to be recompiled and replaced by a prior compiler.
+    for (package, local_unit) in compiled_units {
+        match package_version_map.get(package) {
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == CURRENT_COMPILER_VERSION => {
+                units.push((*package, local_unit.clone()));
+                continue;
+            }
+            Some(toolchain_version) if !package_compiled.contains(package) => {
+                let package_root = SourcePackageLayout::try_find_root(&local_unit.source_path)?;
+                download_and_compile(package_root.clone(), toolchain_version, package)?;
+                package_compiled.insert(*package);
+
+                let compiled_unit_paths = vec![package_root.clone()];
+                let compiled_units = find_filenames(&compiled_unit_paths, |path| {
+                    extension_equals(path, MOVE_COMPILED_EXTENSION)
+                })?;
+                let build_path = package_root.clone().join("build").join(package.as_str());
+                // If we compiled units with a previous compiler, add *all* those units at once, so that
+                // we don't need to resolve them by path one-by-one.
+                for bytecode_path in compiled_units {
+                    info!("bytecode path {bytecode_path}, {package}");
+                    let local_unit =
+                        decode_bytecode_file(build_path.clone(), package, &bytecode_path)?;
+                    units.push((*package, local_unit))
+                }
+                continue;
+            }
+            Some(_) => {
+                // If a toolchain version for a previous compiler exists, we
+                // already added all compiled units to deps_compiled_units.
+                continue;
+            }
+            None => bail!("Expected a version for {package} but none found"),
+        }
+    }
+    Ok(units)
+}
+
+fn download_and_compile(
+    root: PathBuf,
+    ToolchainVersion {
+        compiler_version,
+        edition,
+        flavor,
+    }: &ToolchainVersion,
+    dep_name: &Symbol,
+) -> anyhow::Result<()> {
+    let binaries_path = &*MOVE_HOME; // E.g., ~/.move/binaries
+    let mut dest_dir = PathBuf::from(binaries_path);
+    dest_dir = dest_dir.join("binaries");
+    let dest_version = dest_dir.join(compiler_version.clone());
+    let platform = detect_platform()?;
```

Comment:
This test means that if you're running on a platform that isn't supported by the auto-downloader, you're hosed. Ideally, we would allow people in that situation to sideload a binary at the appropriate location to make progress.  This would require making the following changes:

- Using a platform-agnostic name for the local version of the binary (this should be fine, because each local path only needs to serve one platform -- the one it's on).
- Delaying the `detect_platform()` test to after the check whether `dest_binary.exists()`.
- Updating the error message from `detect_platform()` to explain that sideloading is possible.

Reply:
bed6b8c31e481f6d918e0f1a26c12ff7a97d079b (I had to rebase, so this commit is detached, but it is accurate still)